src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -119,45 +122,12 @@ public PartitionIterator filterReplicaFilteringProtection(PartitionIterator full
     @Override
     public UnfilteredPartitionIterator search(ReadExecutionController executionController) throws RequestTimeoutException
     {
-        // VSTODO see about switching to use an order op instead of ann
-        Supplier<ResultRetriever> queryIndexes = () -> new ResultRetriever(analyze(), analyzeFilter(), controller, executionController, queryContext, command.isTopK());
         if (!command.isTopK())
-            return queryIndexes.get();
-
-        // If there are shadowed primary keys, we have to at least query twice.
-        // First time to find out there are shadowed keys, second time to find out there are no more shadow keys.
-        int loopsCount = 1;
-        final long startShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
-        final var exactLimit = controller.getExactLimit();
-        while (true)
-        {
-            queryContext.addShadowedKeysLoopCount(1L);
-            long lastShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
-            ResultRetriever result = queryIndexes.get();
-            queryContext.resetRowsMatched();
-            UnfilteredPartitionIterator topK = (UnfilteredPartitionIterator) new VectorTopKProcessor(command).filter(result);
-            long currentShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
-            long newShadowedKeysCount = currentShadowedKeysCount - lastShadowedKeysCount;
-            logger.debug("Shadow loop iteration {}: rows matched: {}, keys shadowed: {}",
-                         queryContext.shadowedKeysLoopCount(),
-                         queryContext.rowsMatched(),
-                         currentShadowedKeysCount);
-            // Stop if no new shadowed keys found or if we already got enough rows
-            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.rowsMatched())
-            {
-                cfs.metric.incShadowedKeys(loopsCount, currentShadowedKeysCount - startShadowedKeysCount);
-                if (loopsCount > 1)
-                {
-                    if (newShadowedKeysCount == 0)
-                        Tracing.trace("No new shadowed keys after query loop {}", loopsCount);
-                    else if (exactLimit <= queryContext.rowsMatched())
-                        Tracing.trace("Got enough rows after query loop {}", loopsCount);
-                }
-                return topK;
-            }
-            loopsCount++;
-            Tracing.trace("Found {} new shadowed keys, rerunning query (loop {})", newShadowedKeysCount, loopsCount);
-        }
+            return new ResultRetriever(analyze(), analyzeFilter(), controller, executionController, queryContext, command.isTopK());
+
+        var result = new ScoreOrderedResultRetriever(analyzeOrderedQuery(), analyzeFilter(), controller, executionController, queryContext, command.isTopK());
```

Comment:
Do we still need `command.isTopK()`? Looks like we have two separate control flows here so `ResultRetriever` won't be used for topK search and `ScoreOrderedResultRetriever` is used exclusively for top K seach. Could as well replace it with `true` and `false`, or better just remove the param.

Reply:
Great point, I'll remove it.